Add support to build psa-arch-test protected storage test binary

simonqhughes opened the pull request:
This PR represents the submission of work-in-progress for building the psa-arch-tests protected storage test binary. It includes:
- A repo manifest for creating a workspace containing projects needed to build the test binary.
- A makefile for building the test binary.
- mbed-crypto patches for patching a specific version so the mbed-crypto tests can be leveraged for testing psa_trusted_storage.

This work was carried out under IOTMBL-2532.

simonqhughes committed:
Fix psa_trusted_storage_linux for psa-arch-tests.

The following provides more on this commit:
- The psa-arch-tests requires implementations for the optional extended
  functions psa_ps_create(), psa_ps_set_extended() and psa_ps_get_supported().
  This commit provides stub implementations.
- Some comments are provided to functions.

simonqhughes committed:
Add mbed-crypto patches to support PSA storage testing.

The following provides more information on this commit:
- PSA storage uses the PSA Compliance for testing, but this was unavailable for
  testing during the development of the first implementation. To overcome this
  deficit, mbed-crypto tests were repurposed for PSA storage testing.
- Furthermore, PSA Compliance testing from psa-arch-tests will not be able to
  provide white box testing for PSA storage and so remains a need for
  a project specific test suite.
- While the current mbed-crypto test framework has drawbacks, it's provides
  some additional test coverage prior to a better solution being put in place.
  This will be extended to cover new functionality.

simonqhughes committed:
Add first version of psa-storage-test-manifest.xml.

- psa-storage-test-manifest.xml is a repo manifest file for creating a
  workspace. The workspace is used to build the project test binaries
  using the PSA Compliance testing framework psa-arch-tests, and
  mbed-crypto.

simonqhughes committed:
Add test/Makefile for building test binaries.

The following provides more information on this commit:
- This makefile builds psa-arch-tests-ps, the psa-arch-tests protected storage
  test binary.
- The boiler-plate main.c containing the call to val_entry() has been added to
  the psa-arch-tests tgt_dev_apis_stdc/nspe/common directory. main.o is
  to linked with the psa-arch-tests generated libraries and the libraries under
  test to create the test binary.
- The updated Makefile uses this main.c rather than one from alternative
  repository.
- The makefile also builds the psa-storage tests derived from mbed-crypto.

simonqhughes committed:
Fix lib/Makefile symlink creation to include -f option.

simonqhughes commented:
@jjohnson-arm Please could you take a look at this PR for iotmbl-2532? There is a follow-on PR for a pinned manifest that can only be created after this PR has been merged.

simonqhughes commented:
@jainvikas8 Hi, would you be able to review this PR for me? There are instructions in the test/Makefile 
 for how to build ps-arch-test-ps here: https://github.com/ARMmbed/psa_trusted_storage_linux/pull/9/files#diff-0d8f5a336dfc7f93d0f8ebb4026e4d46

You need to install the repo tool. Let me know if you need help with that.

simonqhughes committed:
Rebase psa-arch-tests sdh_iotmbl_2478 over master and rename to sdh_iotmbl_2532.

The following provides more information on this commit:
- The rebase operation has dropped vikas's tgt_dev_apis_stdc changes,
  which have now been upstreamed.

simonqhughes committed:
Fix library and tests to use valid uid values (i.e. not use zero).

The following provides more information on this commit:
- Fix common_storage.c psa_ps_set()/get_info()/remove() implementations so
  uid=0 causes an error to be returned.
- Fix psa-storage-example-app app/main.c test data so as not use included uid=0
  values.

simonqhughes committed:
Add mbed-crypto test_suite_psa_its patch to fix test not to use uid=0.

The following provides more information on this commit:
- The patch fixes the mbed-crypto test_suite_psa_its.data file so as not
  to use the uid=0 value. The test can then be used to test PSA storage.

simonqhughes committed:
Fix test/Makefile mbed-crypto patching and testing target dependencies.

The following provides more information on this commit:
- Fix Makefile to apply new mbed-crypto patches. The patch fixes mbed-crypto
  tests so as not to use uid=0 where required.
- Fix Makefile test target dependencies to depend on the target that generates
  the binary. This ensures that the test binary has been built before the
  Makefile tried to run the binary.
- Standardise on the use of $(VAR1) rather thatn ${VAR1}.
- Replace the test-dir target with $(TEST_DIR) target i.e. using the actual
  name of the test directory.
- Improve comments.

jainvikas8 commented:
@simonqhughes  I have run the test with `make test-psa-arch-tests-ps` after downloading the repo `make -f proj_mak ws-create`

There are 2 tests failing (406 and 408)
* If we could investigate regarding the failure and raise issue on [psa-arch-tests](https://github.com/ARM-software/psa-arch-tests) or fix as part of the PR, that would be good.

simonqhughes commented:
As requested, issue for Test 406 created here:

https://github.com/ARM-software/psa-arch-tests/issues/125

simonqhughes committed:
Fix inc/config.h to more gracefully handle user supplied symbol definitions.

This fix was requested by the Pelion Cloud Client team. By placing the
inclusion of an optional user supplied config file at the top of config.h
rather than at the bottom, the user supplied config symbols can be defined
without first having to #undef the symbol.

simonqhughes committed:
Fix psa-arch-tests protected storage test 403 for storage exhaustion.

The following provides more information on this commit:
- The storage exhaustion test seeks to create sufficient objects to exhaust
  available storage.
- Two new inc/config.h parameters are introduced to configure system resource
  made available to PS storage:
    - PSA_STORAGE_FILE_MAX. This is the maximum number of file object that can
      be created.
    - PSA_STORAGE_MAX_SIZE. This is the total size of all file obects in bytes.
- The implementation returns an insufficient storage error when either of limits
  is exceeded.

simonqhughes committed:
Fix test/Makefile to optionally output psa-arch-tests cmake make trace.

simonqhughes commented:
@jainvikas8 I've put additional changes on this PR to fix the psa-arch-tests-ps test failures. Test 406 and 408 are now working. Please could you review and hopefully approve? Thanks.

jainvikas8 commented on lib/common_storage.h:
```diff
@@ -10,6 +10,10 @@
 #include "psa/storage_common.h"
 
+/* IHI0087-PS_-Storage_API_1.0.0.pdf specifies 0 to be an invalid uid value
+ * and should be treated as a nerror */
```
***as an error***

simonqhughes committed:
Fix typing errors and code style.
